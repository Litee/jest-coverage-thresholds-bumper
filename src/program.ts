import fs from "fs";
import { DEFAULT_MARGIN, DEFAULT_PATH_TO_COVERAGE_SUMMARY as DEFAULT_PATH_TO_COVERAGE_SUMMARY_FILE } from "./constants";
import { CoverageThresholdsAdapter, createCoverageThresholdsAdapter } from "./coverage-thresholds-adapters";
import { findFileWithJestConfig } from "./jest-config-file-finder";
import { getLogger } from "./logger";
import { resolveFile } from "./utils";
// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageJson = require("../package.json");

const loadCoverageSummary = (coverageSummaryPath: string, options: TypedOptions): CoverageSummary => {
    const filePath = resolveFile(process.cwd(), coverageSummaryPath);
    getLogger(options).info(`Jest coverage summary file: ${filePath}`);
    return require(filePath);
};

export type ThresholdType = "lines" | "statements" | "branches" | "functions";

interface CoverageSummary {
    total: Record<ThresholdType, { pct: number | "Unknown" }>;
}

const updateThresholds = (coverageSummaryFileAsObject: CoverageSummary, coverageThresholdAdapter: CoverageThresholdsAdapter, options: TypedOptions): void => {
    const checkAndUpdateThreshold = (thresholdType: ThresholdType) => {
        const oldValue = coverageThresholdAdapter.getThresholdValue(thresholdType);
        const newValue = coverageSummaryFileAsObject.total[thresholdType].pct;
        if (newValue === "Unknown") {
            getLogger(options).info(`No coverage information for "${thresholdType}" threshold.`);
        } else if (oldValue == null) {
            getLogger(options).info(`Coverage threshold type ${thresholdType} is not defined. Skipping.`);
        } else {
            if (typeof oldValue === "number" && oldValue < 0) {
                // TODO Bump down uncovered lines too
                getLogger(options).info(`Ignoring "${thresholdType}" coverage threshold because it is not a positive number.`);
            } else {
                // Ignore negative values, they specify the maximum number of uncovered lines
                if (oldValue === "Unknown" || (typeof oldValue === "number" && oldValue + options.margin < newValue)) {
                    getLogger(options).info(`Bumping up "${thresholdType}" coverage threshold from ${oldValue} to ${newValue}.`);
                    coverageThresholdAdapter.setThresholdValue(thresholdType, newValue);
                } else {
                    const oldValueDescription = options.margin > 0 ? `current value ${oldValue} + margin ${options.margin}` : `current value ${oldValue}`;
                    getLogger(options).info(
                        `Ignoring "${thresholdType}" coverage threshold because new value ${newValue} is less or equal to ${oldValueDescription}.`
                    );
                }
            }
        }
    };
    // return true is nothing updated
    checkAndUpdateThreshold("lines");
    checkAndUpdateThreshold("statements");
    checkAndUpdateThreshold("branches");
    checkAndUpdateThreshold("functions");
};

const getCoverageSummaryFilePath = (coverageSummaryPath: unknown, options: TypedOptions): string => {
    if (coverageSummaryPath == null) {
        // Try to detect coverageDirectory path
        try {
            const { filePath } = findFileWithJestConfig();
            const configContent = fs.readFileSync(filePath, { encoding: "utf8" });
            const match = configContent.match(`(?:"|')?coverageDirectory(?:"|')?\\s*:\\s*(?:"|')(.*)(?:"|')`);
            if (match && match.length > 1) {
                const coverageDirectory = match[1].trim();
                getLogger(options).info(`Detected custom coverageDirectory value in Jest configuration: ${coverageDirectory}`);
                return coverageDirectory + "/coverage-summary.json";
            }
        } catch (err) {
            getLogger(options).warn("Failed to detect coverageDirectory value, using the default one.");
        }
        return DEFAULT_PATH_TO_COVERAGE_SUMMARY_FILE;
    }
    return coverageSummaryPath as string;
};

export interface TypedOptions {
    dryRun: boolean;
    silent: boolean;
    margin: number;
}

export const execute = (argv: Record<string, unknown>): void => {
    if (!argv.version) {
        const options: TypedOptions = {
            dryRun: (argv.dryRun as boolean) ?? false,
            silent: (argv.silent as boolean) ?? false,
            margin: (argv.margin as number) ?? DEFAULT_MARGIN,
        };
        getLogger(options).info(`Running Jest Coverage Thresholds Bumper v${packageJson.version}...`);
        const coverageSummaryFilePath = getCoverageSummaryFilePath(argv.coverageSummaryPath, options);
        const coverageSummaryFileAsObject = loadCoverageSummary(coverageSummaryFilePath, options);
        const coverageThresholdManager = createCoverageThresholdsAdapter(options);
        getLogger(options).info("Jest configuration and code coverage data loaded. Analyzing...");
        updateThresholds(coverageSummaryFileAsObject, coverageThresholdManager, options);
        coverageThresholdManager.saveIfDirty(options);
        getLogger(options).info("Jest Coverage Thresholds Bumper has finished its work!");
    }
};
