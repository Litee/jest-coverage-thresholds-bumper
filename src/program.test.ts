import "jest";
import { execute } from "./program";
import fs from "fs";

jest.mock("fs");
jest.mock("./utils.ts", () => ({
    resolveFile: (baseDir: string, fileDir: string) => fileDir,
}));

describe("when running execute() function", () => {
    const coverageSummaryFileAsObject = {
        total: {
            lines: { total: 1, covered: 0, skipped: 0, pct: 12 },
            statements: { total: 1, covered: 0, skipped: 0, pct: 23.4 },
            functions: { total: 0, covered: 0, skipped: 0, pct: 34.56 },
            branches: { total: 4, covered: 0, skipped: 0, pct: 45.678 },
        },
    } as const;

    // Some values are higher then values in summary, some values are negative - should not be changed, "statements" value is missing - should be ignored
    const jestConfigJsonFileAsObject1 = {
        coverageThreshold: {
            global: {
                branches: 90,
                functions: -10,
                lines: 10,
            },
        },
    } as const;
    const jestConfigJsonFileAsObject2 = {
        coverageThreshold: {
            global: {
                branches: 1,
                functions: 2.3,
                lines: 3.45,
            },
        },
    } as const;

    beforeEach(() => {
        jest.resetModules();
        fs.writeFileSync = jest.fn();
    });

    describe("and input files exist", () => {
        beforeEach(() => {
            jest.mock(
                "./coverage/coverage-summary.json",
                () => {
                    return coverageSummaryFileAsObject;
                },
                { virtual: true }
            );
        });

        describe.each([
            {
                packageJsonFileAsObject: {
                    name: "jest-coverage-thresholds-bumper-test",
                    version: "0.0.1",
                    jest: JSON.parse(JSON.stringify(jestConfigJsonFileAsObject1)),
                },
            },
            {
                packageJsonFileAsObject: {
                    name: "jest-coverage-thresholds-bumper-test",
                    version: "0.0.1",
                    jest: JSON.parse(JSON.stringify(jestConfigJsonFileAsObject2)),
                },
            },
        ])("and thresholds file has package.json format", ({ packageJsonFileAsObject }) => {
            describe("and jest section exists", () => {
                beforeEach(() => {
                    jest.mock(
                        "./package.json",
                        () => {
                            return packageJsonFileAsObject;
                        },
                        { virtual: true }
                    );
                });

                it("should update coverage thresholds", () => {
                    execute({ dryRun: false });
                    expect(fs.writeFileSync).toBeCalled();
                    expect(fs.writeFileSync).toMatchSnapshot();
                });
            });

            describe("but jest section does not exist", () => {
                beforeEach(() => {
                    const { jest: _, ...packageJsonFileAsObjectWithoutJestSection } = packageJsonFileAsObject;
                    jest.mock(
                        "./package.json",
                        () => {
                            return packageJsonFileAsObjectWithoutJestSection;
                        },
                        { virtual: true }
                    );
                });

                it("should fail", () => {
                    expect(() => execute({ dryRun: false })).toThrowErrorMatchingSnapshot();
                });
            });
        });

        describe.each([
            {
                jestConfigJsonFileAsObject: JSON.parse(JSON.stringify(jestConfigJsonFileAsObject1)),
            },
            {
                jestConfigJsonFileAsObject: JSON.parse(JSON.stringify(jestConfigJsonFileAsObject2)),
            },
        ])("and thresholds file has jest.config.json format", ({ jestConfigJsonFileAsObject }) => {
            beforeEach(() => {
                fs.existsSync = jest.fn((path) => {
                    console.debug(path);
                    return path === "./jest.config.json";
                });
            });

            describe("and coverageThresholds section exists", () => {
                beforeEach(() => {
                    jest.mock(
                        "./jest.config.json",
                        () => {
                            return jestConfigJsonFileAsObject;
                        },
                        { virtual: true }
                    );
                });

                it("should update coverage thresholds", () => {
                    execute({ dryRun: false });
                    expect(fs.writeFileSync).toBeCalled();
                    expect(fs.writeFileSync).toMatchSnapshot();
                });
            });

            describe("but coverageThresholds section does not exist", () => {
                beforeEach(() => {
                    jest.mock(
                        "./jest.config.json",
                        () => {
                            return {};
                        },
                        { virtual: true }
                    );
                });

                it("should not update the file", () => {
                    execute({ dryRun: false });
                    expect(fs.writeFileSync).not.toBeCalled();
                });
            });
        });

        describe.each([
            {
                jestConfigJsFile: `module.exports = {
                    coverageThreshold: {
                        global: {
                            lines: 10,
                        },
                    },
                }` as any,
                margin: 0,
            },
            {
                jestConfigJsFile: `module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 10.0,
                            functions: 1.12,
                            lines: 0.987,
                        },
                    },
                }` as any,
                margin: 0,
            },
            {
                jestConfigJsFile: `module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 40,
                            functions: 30,
                            lines: 10,
                        },
                    },
                }` as any,
                margin: 5,
            },
        ])("and thresholds file has jest.config.js format", ({ jestConfigJsFile, margin }) => {
            beforeEach(() => {
                fs.existsSync = jest.fn((path) => {
                    console.debug(path);
                    return path === "./jest.config.js";
                });
                fs.readFileSync = jest.fn(() => {
                    return jestConfigJsFile;
                });
            });

            it("should update coverage thresholds", () => {
                execute({ margin, dryRun: false });
                expect(fs.writeFileSync).toBeCalled();
                expect(fs.writeFileSync).toMatchSnapshot();
            });
        });

        describe.each([
            {
                jestConfigTsFile: `module.exports = {
                    coverageThreshold: {
                        global: {
                            lines: 10,
                        },
                    },
                }` as any,
                margin: 0,
            },
            {
                jestConfigTsFile: `module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 10.0,
                            functions: 1.12,
                            lines: 0.987,
                        },
                    },
                }` as any,
                margin: 0,
            },
            {
                jestConfigTsFile: `module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 40,
                            functions: 30,
                            lines: 10,
                        },
                    },
                }` as any,
                margin: 5,
            },
        ])("and thresholds file has jest.config.ts format", ({ jestConfigTsFile, margin }) => {
            beforeEach(() => {
                fs.existsSync = jest.fn((path) => {
                    console.debug(path);
                    return path === "./jest.config.ts";
                });
                fs.readFileSync = jest.fn(() => {
                    return jestConfigTsFile;
                });
            });

            it("should update coverage thresholds", () => {
                execute({ margin, dryRun: false });
                expect(fs.writeFileSync).toBeCalled();
                expect(fs.writeFileSync).toMatchSnapshot();
            });
        });
    });
});
