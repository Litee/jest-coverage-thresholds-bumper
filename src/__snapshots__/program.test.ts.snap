// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when running execute() function and input files exist and thresholds file has jest.config.js format should update coverage thresholds 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.js",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            lines: 12,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.js format should update coverage thresholds 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.js",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 45.678,
                            functions: 34.56,
                            lines: 12,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.js format should update coverage thresholds 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.js",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 45.678,
                            functions: 30,
                            lines: 10,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.js format specified as argument should update coverage thresholds 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./custom/jest.custom.config.js",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            lines: 12,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.js format specified as argument should update coverage thresholds 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./custom/jest.custom.config.js",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 45.678,
                            functions: 34.56,
                            lines: 12,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.js format specified as argument should update coverage thresholds 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./custom/jest.custom.config.js",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 45.678,
                            functions: 30,
                            lines: 10,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.json format and coverageThresholds section exists should update coverage thresholds 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.json",
      "{
  \\"coverageThreshold\\": {
    \\"global\\": {
      \\"branches\\": 90,
      \\"functions\\": -10,
      \\"lines\\": 12
    }
  }
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.json format and coverageThresholds section exists should update coverage thresholds 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.json",
      "{
  \\"coverageThreshold\\": {
    \\"global\\": {
      \\"branches\\": 45.678,
      \\"functions\\": 34.56,
      \\"lines\\": 12
    }
  }
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.json format specified as an argument and coverageThresholds section exists should update coverage thresholds 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./custom/jest.custom.config.json",
      "{
  \\"coverageThreshold\\": {
    \\"global\\": {
      \\"branches\\": 90,
      \\"functions\\": -10,
      \\"lines\\": 12
    }
  }
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.json format specified as an argument and coverageThresholds section exists should update coverage thresholds 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./custom/jest.custom.config.json",
      "{
  \\"coverageThreshold\\": {
    \\"global\\": {
      \\"branches\\": 45.678,
      \\"functions\\": 34.56,
      \\"lines\\": 12
    }
  }
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.ts format should update coverage thresholds 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.ts",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            lines: 12,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.ts format should update coverage thresholds 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.ts",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 45.678,
                            functions: 34.56,
                            lines: 12,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has jest.config.ts format should update coverage thresholds 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./jest.config.ts",
      "module.exports = {
                    coverageThreshold: {
                        global: {
                            branches: 45.678,
                            functions: 30,
                            lines: 10,
                        },
                    },
                }",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has package.json format and jest section exists should update coverage thresholds 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./package.json",
      "{
  \\"name\\": \\"jest-coverage-thresholds-bumper-test\\",
  \\"version\\": \\"0.0.1\\",
  \\"jest\\": {
    \\"coverageThreshold\\": {
      \\"global\\": {
        \\"branches\\": 90,
        \\"functions\\": -10,
        \\"lines\\": 12
      }
    }
  }
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has package.json format and jest section exists should update coverage thresholds 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "./package.json",
      "{
  \\"name\\": \\"jest-coverage-thresholds-bumper-test\\",
  \\"version\\": \\"0.0.1\\",
  \\"jest\\": {
    \\"coverageThreshold\\": {
      \\"global\\": {
        \\"branches\\": 45.678,
        \\"functions\\": 34.56,
        \\"lines\\": 12
      }
    }
  }
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`when running execute() function and input files exist and thresholds file has package.json format but jest section does not exist should fail 1`] = `"package.json file has no 'jest' section!"`;

exports[`when running execute() function and input files exist and thresholds file has package.json format but jest section does not exist should fail 2`] = `"package.json file has no 'jest' section!"`;
